### ISSUE TITLE
GPU available memory reported incorrectly under Windows 7 64 bit OS

### ISSUE THREAD

**Issue by brendanruff** [Theano/Theano#5137]
Apologies if this has already been sorted but I could not find a solution.

I have Windows 7 64 bit OS, I run 64 bit Python, I use VS2015 and python tools 2.2 (it is integrated). I have TitanX Pascal 12GB (dedicated VRAM).

Theano only accesses 4GB and also calling cuda_ndarray.mem_info() returns a -ve number. This all suggests that theano is accessing the 32 bit API in CUDA on Windows to get the memory info rather than the 64 bit API. If there is a workaround or solution or if I am just being ignorant (quite likely) please lead me on the right path and forgive my errors.

**Comment by nouiz** [comment]
We don't use a specific 32 vs 64 bit api for cuda. What do you mean by
mem_info() reutrn a -ve number? Can you give the exact output?
Di you installed the CUDA 32 bit driver? If so, install the 64 bit driver.
We always suppose all the stack is 32 bit or 64 bit. We don't support
having a mixed environment on that parameter.

Make sure to use Theano dev version.

**Comment by brendanruff** [comment]
Hi Frederic

 

Thanks for your reply. This is a new PC on Windows 64 and the latest CUDA driver. Also I have another windows pc with exactly the same issue. Tomorrow I will delve into the Theano source code to try and find out what is happening, but clearly this is a 32 bit issue. Curiously GPU-Z reports 12GB memory but only 4GB dedicated (ie VRAM) memory. I also see this on the other PC. I contacted nVidia and they suggested that this is due to accessing a 32bit interface to CUDA. So, I am thinking somewhere theano is accessing the 32 bit API. Apparently the 32bit API exists on a 64bit installation, so nVidia strongly suggest this is the app (ie theano) accessing the wrong API.

 

By the way, if you are interested I just proved a simpler way to train deep CNNs without batch normalisation, ie based on single samples. It converges faster and to a better solution (eg with Citysacapes and semantic segmentation and with Segnet). I worked on a new theory and understanding of what batch normalisation is doing, and it is wrong but parts of what it does is correct. Also I removed weight decay which helps BN but ultimately is destructive to training models. I replace it with another regulariser designed to encourage a certain absolute range to the last conv layer’s output. I intend to publish this in November on Arxiv but if you are interested to try it out beforehand then please let me know and I will release the technical details to you.

 

Best regards

 

Brendan Ruff

**Comment by lamblin** [comment]
This is related to the old GPU back-end, which we removed.
If you encounter a similar issue with the new back-end (the one using libgpuarray), please open a new issue.
